render-and-diff-helm/action.yml

```diff
@@ -0,0 +1,179 @@
+          if [ -d "${{ matrix.chart }}" ]; then
+            helm dependency build "${{ matrix.chart }}"
+            values_files="${{ matrix.chart }}"/values-*
+            for values_file in $(basename -a $values_files); do
+              helm template "${{ matrix.chart }}" -f "${{ matrix.chart }}/values.yaml" -f "${{ matrix.chart }}/${values_file}" --output-dir "shared/head-charts/${{ matrix.chart }}/${values_file}"
+            done
+          fi
+          echo sanitized_name=$(echo "${{ matrix.chart }}" | sed 's/\//-/g') >> $GITHUB_OUTPUT
+      - name: upload artifact
+        uses: actions/upload-artifact@v4
+        with:
+          name: "shared-head-${{ steps.render_head.outputs.sanitized_name }}"
+          path: "shared"
+
+  render_base_ref_charts:
+    runs-on: ubuntu-latest
+    needs: get_changed_helm_charts
+    strategy:
+      matrix:
+        chart: ${{ fromJSON(needs.get_changed_helm_charts.outputs.matrix_charts) }}
+    steps:
+      - name: checkout repository
+        uses: actions/checkout@v4
+
+      - name: setup helm
+        uses: azure/setup-helm@v4.0.0
+
+      - name: render ${{ matrix.chart }} from base ref
+        id: render_base
+        run: |
+          mkdir -p shared/base-charts
+          git fetch origin ${{ github.base_ref }}
+          git checkout  ${{ github.base_ref }} --
+          if [ -d "${{ matrix.chart }}" ]; then
+            helm dependency build "${{ matrix.chart }}"
+            values_files="${{ matrix.chart }}"/values-*
+            for values_file in $(basename -a $values_files); do
+              helm template "${{ matrix.chart }}" -f "${{ matrix.chart }}/values.yaml" -f "${{ matrix.chart }}/${values_file}" --output-dir "shared/base-charts/${{ matrix.chart }}/${values_file}"
+            done
+          fi
+          echo sanitized_name=$(echo "${{ matrix.chart }}" | sed 's/\//-/g') >> $GITHUB_OUTPUT
+      - name: upload artifact
+        uses: actions/upload-artifact@v4
+        with:
+          name: "shared-base-${{ steps.render_base.outputs.sanitized_name }}"
+          path: "shared"
+
+  diff_helm_charts:
+    runs-on: ubuntu-latest
+    needs:
+      - get_changed_helm_charts
+      - render_base_ref_charts
+      - render_head_ref_charts
+    steps:
+      - name: setup helm
+        uses: azure/setup-helm@v4.0.0
+
+      - name: download artifacts
+        uses: actions/download-artifact@v4
+        with:
+          pattern: shared-*
+          merge-multiple: true
+          path: "shared"
+
+      - name: diff helm charts
+        id: diff_helm_charts
+        run: |
+          for chart in ${{ needs.get_changed_helm_charts.outputs.charts }}; do
+            chart_diff_output=$(diff -r "shared/base-charts/${chart}" "shared/head-charts/${chart}" || true)
+            if [ -n "$chart_diff_output" ]; then
+              echo -e "Changes found in chart: ${chart}\n$(diff -ru shared/base-charts/${chart} shared/head-charts/${chart})\n" >> diff.log
```

Comment:
The unified context for diff shows the new changes with a time of the beginning of epoch. Although I am not sure how to make this any better, just a note since it could be confusing. I looked through diff man page for other alternates, but nothing comes to mind. I think for now it's fine the way it is.

Reply:
We could potentially use `git diff -u --no-index` instead which will not have timestamps